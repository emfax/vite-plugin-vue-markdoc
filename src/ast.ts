import {
  ArrayExpression,
  ArrowFunctionExpression,
  ExportNamedDeclaration,
  Expression,
  Literal,
  ObjectExpression,
  Program,
  Property,
} from "estree";

export function addObjectProperty(obj: ObjectExpression, key: string, value: Expression): void {
  obj.properties.push(newProperty(key, value));
}

export function newArrayExpression(): ArrayExpression {
  return {
    type: "ArrayExpression",
    elements: [],
  };
}

export function newExportNamedDeclaration(name: string, init: Expression): ExportNamedDeclaration {
  return {
    type: "ExportNamedDeclaration",
    declaration: {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          "id": {
            type: "Identifier",
            name,
          },
          init,
        }
      ],
      kind: "const",
    },
    specifiers: [],
  };
}

export function newImportFunction(path: string): ArrowFunctionExpression {
  return {
    type: "ArrowFunctionExpression",
    expression: true,
    generator: false,
    async: false,
    params: [],
    body: {
      type: "ImportExpression",
      source: {
        type: "Literal",
        value: path,
      },
    }
  };
}

export function newLiteral(value?: string | number): Literal {
  return {
    type: "Literal",
    value,
  };
}

export function newObjectExpression(): ObjectExpression {
  return { type: "ObjectExpression", properties: [] };
}

export function arrayExpressionPush(arr: ArrayExpression, element: Expression) {
  arr.elements.push(element);
}

export function newProgram(...body: ExportNamedDeclaration[]): Program {
  return {
    type: "Program",
    body,
    sourceType: "module",
  };
}

export function newProperty(key: string, value: Expression): Property {
  return {
    type: "Property",
    method: false,
    shorthand: false,
    computed: false,
    kind: "init",
    key: {
      type: "Identifier",
      name: key,
    },
    value,
  };
}
