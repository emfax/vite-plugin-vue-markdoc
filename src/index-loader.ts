import { dirname, join, resolve } from "node:path";
import Markdoc from "@markdoc/markdoc";
import type { Node } from "@markdoc/markdoc";
import { Ok, Err, Result } from "./common";
import yaml from "js-yaml";
import {
  ArrayExpression,
  ObjectExpression,
} from "estree";
import { generate } from "astring";
import {
  addObjectProperty,
  newArrayExpression,
  newExportNamedDeclaration,
  newImportFunction,
  newLiteral,
  newObjectExpression,
  newProgram,
  newProperty,
} from "./ast";

/**
 * Decrement the last item in an array of integers and return the new value.
 * Return null if list is empty.
 * 
 * @param stack 
 */
function decrementTail(stack: number[]): number | null {
  if (stack.length === 0) return null;
  const result = --stack[stack.length - 1];

  return result;
}

function foldToc(arrays: ArrayExpression[]): void {
  if (arrays.length <= 1) return;

  const children = arrays.pop();
  const parentArray = arrays[arrays.length - 1];
  const parentObject = parentArray.elements.pop() as ObjectExpression;
  addObjectProperty(parentObject, "children", children);
  parentArray.elements.push(parentObject);
}

function generateModule(routes: ArrayExpression, toc: ArrayExpression): string {
  return generate(
    newProgram(
      newExportNamedDeclaration("routes", routes),
      newExportNamedDeclaration("toc", toc),
    )
  );
}

function last<T>(arr: T[]): T | null {
  if (arr.length === 0) return null;
  
  return arr[arr.length - 1];
}

// function newTocItem(text: string, path?: string): ObjectExpression {
//   const properties = [
//     newProperty("text", { type: "Literal", value: text }),
//   ];

//   if (path) {
//     properties.push(newProperty("path", { type: "Literal", value: path }));
//   }

//   return {
//     type: "ObjectExpression",
//     properties,
//   };
// }

function newRouteRecord(path: string, url: string): ObjectExpression {
  return {
    type: "ObjectExpression",
    properties: [
      newProperty("path", { type: "Literal", value: path }),
      newProperty("component", newImportFunction(url)),
    ],
  };
}

function push(arr: ArrayExpression, el: ObjectExpression) {
  arr.elements.push(el);
}

export function indexLoader(source: string, id: string): Result<string | null, string> {
  const ast = Markdoc.parse(source);

  const frontmatter = yaml.load(ast.attributes.frontmatter);

  const list = ast.children[0];
  if (list.type !== "list") {
    return null;
  };

  const routes = newArrayExpression();
  const toc = newArrayExpression();

  nest(list, routes, toc);

  return Ok(generateModule(routes, toc));

  function nest(list: Node,  routes: ArrayExpression, toc: ArrayExpression): Result<undefined, string> {

    for (const listItem of list.children) {
      const tocItem = newObjectExpression();

      for (let node of listItem.children) {
        if (node.type === "inline") {
          node = node.children[0]; // node.type === "link"
          
          if (node.attributes.href) {
            const path = resolve(
              frontmatter.baseUrl || "",
              node.attributes.href.replace(/(readme)?\.md$/i, "")
            );
            const url = resolve(dirname(id), node.attributes.href);

            addObjectProperty(tocItem, "path", newLiteral(path));

            push(routes, newRouteRecord(path, url));
          }

          node = node.children[0]; // node.type === "text"

          if (!node) {
            return Err("index entries must have text");
          }

          addObjectProperty(tocItem, "text", newLiteral(node.attributes.content));

          continue;
        }

        if (node.type === "list") {
          const children = newArrayExpression();
          const result = nest(node, routes, children);
          if (!result.ok) return result;
          addObjectProperty(tocItem, "children", children);
        }
      }

      push(toc, tocItem);
    }

    return Ok();
  }
}
