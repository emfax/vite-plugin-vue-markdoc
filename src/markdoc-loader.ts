import Markdoc from "@markdoc/markdoc";

export default function* markdocToVue(source: string, options = {}): Generator<string, string, string> {
  const ast = Markdoc.parse(source);

  for (const child of ast.walk()) {
    if (child.type === "image") {
      child.attributes.src = yield child.attributes.src;
    }
  }

  const renderable = Markdoc.transform(ast, options);
  const html = Markdoc.renderers.html(renderable);

  return `<template>${html}</template>`;
}
