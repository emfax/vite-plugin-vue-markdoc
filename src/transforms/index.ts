export * as heading from "./heading";
export * as link from "./link";
