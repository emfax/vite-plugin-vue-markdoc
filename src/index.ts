import Markdoc from "@markdoc/markdoc";
import { generate } from "astring";
import { ArrayExpression, ArrowFunctionExpression, ExportNamedDeclaration, Expression, ObjectExpression, Program, Property } from "estree";
import type { Plugin } from "vite";
import { dirname, join, resolve } from "node:path";
import markdocToVue from "./markdoc-loader";
import { cwd } from "node:process";
import { readFileSync } from "node:fs";
import yaml from "js-yaml";
import { addObjectProperty } from "./ast";
import link from "./transforms/link";

export interface Options {
  markdoc: Record<string, string>;
}

const filter = (id: string) => {
  if (id.endsWith(".md")) return true;
  if (id.endsWith(".md.vue")) return true;
};

const INDEX_RE = /index\.md$/i;

export default function (): Plugin {
  return {
    enforce: "pre",

    name: "vite-plugin-vue-markdoc",

    async resolveId(source, importer) {
      if (source.endsWith(".md") && !INDEX_RE.test(source)) {
        const resolution = await this.resolve(source, importer, { skipSelf: true });
        return `${resolution.id}.vue`;
      }

      return null;
    },

    load(id) {
      if (id.endsWith(".md.vue")) {        
        let path = id.slice(0, -4);
        if (process.env.NODE_ENV === "development") {
          path = join(cwd(), path);
        }

        return readFileSync(path, "utf-8");
      }

      return null;
    },

    async transform(source, id) {
      if (!filter(id)) {
        return null;
      }

      if (id.endsWith(".md.vue")) {
        const iter = markdocToVue(source, { nodes: { link }});
        let curr = iter.next();
        while (!curr.done) {
          const src = curr.value;

          curr = iter.next(resolve(dirname(id), src));
        }

        console.log(curr.value);
        return curr.value;
      }

      const ast = Markdoc.parse(source);

      const frontmatter = yaml.load(ast.attributes.frontmatter);

      const list = ast.children[0];

      if (list.type !== "list") {
        return null;
      };

      const stack = [list];
      const routes = newArrayExpression();
      const toc = [newArrayExpression()];
      for (const child of list.walk()) {
        stack.push(child);

        if (child.type === "list") {
          toc.push(newArrayExpression());
        }

        if (child.type === "link") {
          if (child.children.length === 0) {
            this.error("links in markdown index must have text");
          }
    
          continue;
        }

        if (child.type === "text") {
          const text = stack.pop();
          const link = stack.pop();

          // Pop the `inline` Node
          stack.pop();

          const item = stack.pop();
          let list = stack[stack.length - 1];

          if (!list.children.includes(item)) {
            stack.pop(); // List
            stack.pop(); // Item that wraps list
            list = stack[stack.length - 1];

            if (toc.length > 1) {
              foldToc(toc);
            }
          }

          const path = join(
            frontmatter.baseUrl || "",
            link.attributes.href.replace(/(readme)?\.md$/i, ""),
          );

          push(toc[toc.length - 1], newTocItem(text.attributes.content, path));

          if (link.attributes.href) {
            push(routes, newRouteRecord(path, resolve(dirname(id), link.attributes.href)));
          }
        }
      }

      return generateModule(routes, toc.pop());
    },
  };
}

function newArrayExpression(): ArrayExpression {
  return {
    type: "ArrayExpression",
    elements: [],
  };
}

function newProperty(key: string, value: Expression): Property {
  return {
    type: "Property",
    method: false,
    shorthand: false,
    computed: false,
    kind: "init",
    key: {
      type: "Identifier",
      name: key,
    },
    value,
  };
}

function newImportFunction(path: string): ArrowFunctionExpression {
  return {
    type: "ArrowFunctionExpression",
    expression: true,
    generator: false,
    async: false,
    params: [],
    body: {
      type: "ImportExpression",
      source: {
        type: "Literal",
        value: path,
      },
    }
  };
}

function generateModule(routes: ArrayExpression, toc: ArrayExpression): string {
  return generate(
    newProgram(
      newExportNamedDeclaration("routes", routes),
      newExportNamedDeclaration("toc", toc),
    )
  );
}

function newExportNamedDeclaration(name: string, init: Expression): ExportNamedDeclaration {
  return {
    type: "ExportNamedDeclaration",
    declaration: {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          "id": {
            type: "Identifier",
            name,
          },
          init,
        }
      ],
      kind: "const",
    },
    specifiers: [],
  };
}

function newProgram(...body: ExportNamedDeclaration[]): Program {
  return {
    type: "Program",
    body,
    sourceType: "module",
  };
}

function newRouteRecord(path: string, url: string): ObjectExpression {
  return {
    type: "ObjectExpression",
    properties: [
      newProperty("path", { type: "Literal", value: path }),
      newProperty("component", newImportFunction(url)),
    ],
  };
}

function newTocItem(text: string, path?: string): ObjectExpression {
  const properties = [
    newProperty("text", { type: "Literal", value: text }),
  ];

  if (path) {
    properties.push(newProperty("path", { type: "Literal", value: path }));
  }

  return {
    type: "ObjectExpression",
    properties,
  };
}

function push(arr: ArrayExpression, el: ObjectExpression) {
  arr.elements.push(el);
}

function foldToc(arrays: ArrayExpression[]): void {
  const children = arrays.pop();
  const parentArray = arrays[arrays.length - 1];
  const parentObject = parentArray.elements.pop() as ObjectExpression;
  addObjectProperty(parentObject, "children", children);
  parentArray.elements.push(parentObject);
}
