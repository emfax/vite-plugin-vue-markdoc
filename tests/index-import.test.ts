import { readFile } from "node:fs/promises";
import { describe, expect, test } from "vitest";
import { indexLoader } from "../src/index-loader";

describe('index importer', () => {
  test("tricky index", async () => {
    const md = await readFile("tests/fixtures/tricky-index.md", "utf-8");
    const result = indexLoader(md, "/tricky-index.md");

    if (result.ok) {
      expect(result.value).toMatchSnapshot();
    } else if (result.ok === false) {
      throw new Error(result.error);
    }
  });

  test.skip("weird index", async () => {
    const md = await readFile("tests/fixtures/weird-index.md", "utf-8");
    const result = indexLoader(md, "/weird-index.md");

    if (result.ok) {
      expect(result.value).toMatchSnapshot();
    } else if (result.ok === false) {
      throw new Error(result.error);
    }
  });
});
