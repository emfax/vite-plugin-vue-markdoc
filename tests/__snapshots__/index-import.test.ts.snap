// Vitest Snapshot v1

exports[`index importer > tricky index 1`] = `
"export const routes = [{
  path: \\"/\\",
  component: () => import(\\"/README.md\\")
}, {
  path: \\"/level-one\\",
  component: () => import(\\"/level-one.md\\")
}, {
  path: \\"/level-two\\",
  component: () => import(\\"/level-two.md\\")
}, {
  path: \\"/level-three\\",
  component: () => import(\\"/level-three.md\\")
}, {
  path: \\"/level-four\\",
  component: () => import(\\"/level-four.md\\")
}, {
  path: \\"/again\\",
  component: () => import(\\"/again.md\\")
}];
export const toc = [{
  path: \\"/\\",
  text: \\"Introduction\\"
}, {
  path: \\"/level-one\\",
  text: \\"Level One\\",
  children: [{
    path: \\"/level-two\\",
    text: \\"Level Two\\",
    children: [{
      path: \\"/level-three\\",
      text: \\"Level Three\\",
      children: [{
        path: \\"/level-four\\",
        text: \\"Level Four\\"
      }]
    }]
  }]
}, {
  path: \\"/again\\",
  text: \\"Back to Level One\\"
}];
"
`;
